### PR TITLE
configuration cascades and retire add fields

### PR DESCRIPTION
**Pull request by boutell**

### CODE REVIEW

**Comment by boutell** [comment]
w00t! Ready for review.

It's a lot of changes, mostly to actually do the configuration using the features, but zero in on these files:

doc-type/index.js
piece-type/index.js
widget-type/index.js
moog.js

Happy to walk through it etc.

**Comment by boutell** [comment]
Note that we got the full support for declaring more cascades as needed in various modules.

For now at least, all of the cascades are getting turned into arrays at the end by the module that wanted them, to avoid completely rewriting the schema module today, etc. But that can evolve if we really want it to. The important thing is to support the object cascade at configuration time.

**Comment by abea** [comment]
Screen Shot 2020-09-02 at 9 31 13 AM

Love to see that.

File: lib/moog.js [hunk 1]
```diff
@@ -146,36 +146,53 @@ module.exports = function(options) {
...
-            ...fields.groups[name]
-          })));
+        if (!that[`${cascade}Groups`]) {
+          that[`${cascade}Groups`] = {};
         }
-        for (const key of Object.keys(fields)) {
-          if (!valid.includes(key)) {
-            throw `${clarifyModuleName(step.__meta.name)}: ${key} is not valid inside "fields".\nPossibly you forgot to nest a field in "add".`;
+        // You can have access to options within a function, if you choose to provide one
+        const fields = ((typeof step[cascade]) === 'function') ? step[cascade](that, options) : step[cascade];
```
Comment (abea) [inline lib/moog.js:172]: @boutell How is this specific to the `fields` option on a module?

**Review by falkodev — APPROVED** [review]

File: lib/moog.js [hunk 2]
```diff
@@ -146,42 +146,59 @@ module.exports = function(options) {
...
+        if (!that[`${cascade}Groups`]) {
+          that[`${cascade}Groups`] = {};
         }
-        for (const key of Object.keys(fields)) {
-          if (!valid.includes(key)) {
-            throw `${clarifyModuleName(step.__meta.name)}: ${key} is not valid inside "fields".\nPossibly you forgot to nest a field in "add".`;
+        // You can have access to options within a function, if you choose to provide one
+        const properties = ((typeof step[cascade]) === 'function') ? step[cascade](that, options) : step[cascade];
+        if (properties) {
+          const valid = [ 'add', 'remove', 'group' ];
```
[the remark this answers is not in the document]
Comment (abea) [reply lib/moog.js:174]: There was a brief discussion of that yesterday afternoon (US Eastern Time) to use `group` as the verb since the others are verbs.

**Comment by boutell** [comment]
Ha, this was a discussion yesterday.

It is `group` because `add` and `remove` are both verbs, so `group` is a verb as well. You can read it as a command: "group these things."

**Pull request merged**